contracts/fio.perms/fio.perms.cpp

```diff
@@ -0,0 +1,477 @@
+
+
+            fio_400_assert(permission_name.length()  > 0, "permission_name", permission_name,
+                           "Permission name is invalid", ErrorInvalidPermissionName);
+            //  error if permission name is not the expected name register_address_on_domain.
+            //one permission name is integrated for fip 40, modify this logic for any new permission names
+            //being supported
+            fio_400_assert(useperm.compare(REGISTER_ADDRESS_ON_DOMAIN_PERMISSION_NAME) == 0, "permission_name", permission_name,
+                           "Permission name is invalid", ErrorInvalidPermissionName);
+
+            // error if permission info is not empty.
+            fio_400_assert(permission_info.size()  == 0, "permission_info", permission_info,
+                           "Permission info is invalid", ErrorInvalidPermissionInfo);
+            // error if object name is not * or is not in the domains table
+            fio_400_assert(object_name.size()  > 0, "object_name", object_name,
+                           "Object name is invalid", ErrorInvalidObjectName);
+
+            //verify domain name, and that domain is owned by the actor account.
+            FioAddress fa;
+            getFioAddressStruct(object_name, fa);
+
+            fio_400_assert(fa.domainOnly, "domonlyobject_name", object_name, "Invalid object name",
+                           ErrorInvalidObjectName);
+
+            const uint128_t domainHash = string_to_uint128_hash(fa.fiodomain.c_str());
+            auto domainsbyname = domains.get_index<"byname"_n>();
+            auto domains_iter = domainsbyname.find(domainHash);
+
+            fio_400_assert(domains_iter != domainsbyname.end(), "domnotfound object_name", object_name,
+                           "Invalid object name",
+                           ErrorInvalidObjectName);
+
+            //add 30 days to the domain expiration, this call will work until 30 days past expire.
+            const uint32_t domain_expiration = get_time_plus_seconds(domains_iter->expiration, SECONDS30DAYS);
+
+            const uint32_t present_time = now();
+            fio_400_assert(present_time <= domain_expiration, "domexpired object_name", object_name, "Invalid object name",
+                           ErrorInvalidObjectName);
+
+            //check domain owner is actor
+            fio_400_assert((actor.value == domains_iter->account), "actornotowner object_name", object_name,
+                           "Invalid object name", ErrorInvalidObjectName);
+
+
+            //check grantee exists.
+            fio_400_assert(is_account(grantee_account), "grantee_account", grantee_account.to_string(),
+                           "grantee account is invalid", ErrorInvalidGranteeAccount);
+
+            fio_400_assert((grantee_account.value != actor.value), "grantee_account", grantee_account.to_string(),
+                           "grantee account is invalid", ErrorInvalidGranteeAccount);
```

Comment:
Is this a check that ensures that the actor is not granting themselves permission? If so, we should have a clearer error like "Cannot grant permissions to self"